src/wifi/model/periodic-dmg-wifi-scheduler.cc

```diff
@@ -0,0 +1,539 @@
+      endAlloc = startAlloc + allocation.GetAllocationBlockDuration () + m_guardTime;
+      if (allocation.GetNumberOfBlocks () > 1)
+        {
+          // if the number of blocks allocated is > 1, the allocation is periodic and must be dealt with accordingly
+          uint8_t blocks = allocation.GetNumberOfBlocks ();
+          while (blocks > 0)
+            {
+              UpdateAvailableSlots (startAlloc, endAlloc);
+              startAlloc += allocation.GetAllocationBlockPeriod ();
+              endAlloc += allocation.GetAllocationBlockPeriod ();
+              // AllocationBlockPeriod represents the time between the start of two consecutive time blocks belonging to the same allocation
+              blocks -= 1;
+            }
+        }
+      else
+        {
+          UpdateAvailableSlots (startAlloc, endAlloc);
+        }
+    }
+}
+
+uint32_t
+PeriodicDmgWifiScheduler::GetAllocationDuration (uint32_t minAllocation, uint32_t maxAllocation)
+{
+  NS_LOG_FUNCTION (this << minAllocation << maxAllocation);
+  return maxAllocation;
+}
+
+StatusCode
+PeriodicDmgWifiScheduler::AddNewAllocation (uint8_t sourceAid, const DmgTspecElement &dmgTspec, const DmgAllocationInfo &info)
+{
+  NS_LOG_FUNCTION (this);
+
+  StatusCode status;
+  uint32_t allocDuration;
+
+  if (m_availableSlots.begin () == m_availableSlots.end ())
+    {
+      NS_LOG_DEBUG ("There are no free available slots in the DTI.");
+      status.SetFailure ();
+      return status;
+    }
+
+  if (info.GetAllocationFormat () == ISOCHRONOUS)
+    {
+      allocDuration = GetAllocationDuration (dmgTspec.GetMinimumAllocation (), dmgTspec.GetMaximumAllocation ());
+      if (allocDuration < dmgTspec.GetMinimumAllocation ())
+        {
+          NS_LOG_DEBUG ("Unable to guarantee minimum duration.");
+          status.SetFailure ();
+          return status;
+        }
+    }
+  else if (info.GetAllocationFormat () == ASYNCHRONOUS)
+    {
+      // for asynchronous allocations, the Maximum Allocation field is reserved (IEEE 802.11ad 8.4.2.136)
+      allocDuration = dmgTspec.GetMinimumAllocation ();
+    }
+  else
+    {
+      NS_FATAL_ERROR ("Allocation Format not supported");
+    }
+
+  // Implementation of an admission policy for newly received requests. 
+  uint16_t allocPeriod = dmgTspec.GetAllocationPeriod ();
+  if (allocPeriod != 0)
+    {
+      // Proceed with the allocation of periodic SPs 
+      bool isMultiple = dmgTspec.IsAllocationPeriodMultipleBI ();
+      if (isMultiple)
+        {
+          NS_FATAL_ERROR ("Multiple BI periodicity is not supported yet.");
+        }
+
+      // spInterval is going to be passed to AddAllocationPeriod to specify the 
+      // distance between consecutive periodic SPs
+      uint32_t spInterval = uint32_t (m_biDuration.GetMicroSeconds () / allocPeriod);
+
+      NS_LOG_DEBUG ("Allocation Period " << uint32_t (allocPeriod) << " AllocDuration " << allocDuration << " Multiple " << isMultiple);
```

Comment:
you should be able to print `uint16_t` directly
```suggestion
      NS_LOG_DEBUG ("Allocation Period " << allocPeriod << " AllocDuration " << allocDuration << " Multiple " << isMultiple);
```